File: SCC-strong-connected-components/Tarjan.cpp
#include <iostream>
#include <algorithm>
#include <cstring>
#include <stack>
using namespace std;

const int maxn = 105;
int dfn[maxn], low[maxn];
bool flag[maxn];
int length, cnt = 0, number = 0;
int matrix[105][105]={
        {0,1,1,0,0,0},{0,0,0,1,0,0},{0,0,0,1,1,0},
        {1,0,0,0,0,1},{0,0,0,0,0,1},{0,0,0,0,0,0}
};
stack<int> s;
vector<vector<int>> ans;


vector<vector<int>> tarjan(int u)
{
    dfn[u] = low[u] = ++cnt;
    s.push(u);
    flag[u] = true;
    for (int v = 0; v < length; ++v) {
        if (matrix[u][v]){
            if (!dfn[v]){
                tarjan(v);
                low[u] = min(low[u], low[v]);
            }
            else if(flag[v])
                low[u] = min(low[u], dfn[v]);
        }
    }
    int j;

    if (dfn[u] == low[u]){
        number++;
        vector<int> tmp;
        do {
            j = s.top();
            tmp.push_back(j);
            s.pop();
            flag[j] = false;
        }while (j != u);
        ans.push_back(tmp);
    }
    return ans;
}


void print_ans(vector<vector<int>> ans)
{
    int n = ans.size();
    for (int i = 0; i < n; ++i) {
        int m = ans[i].size();
        for (int j = 0; j < m; ++j) {
            cout<<ans[i][j]<<(j == m-1?"\n":" ");
        }
    }
}


int main()
{
    ios_base::sync_with_stdio(false);
    memset(dfn, 0, sizeof(dfn));
    memset(low, 0, sizeof(low));
    memset(flag, false, sizeof(flag));
    length = 6;
    vector<vector<int>> ans = tarjan(0);
    print_ans(ans);
    return 0;
}

File: SCC-strong-connected-components/tarjan.py
class Tarjan:
    ''' Tarjan algorithm for strong connected component '''
    def __init__(self, graph:'List[List[int]]'):
        self.g = graph
        self.length = len(self.g)
        self.dfn, self.low = [0] * self.length, [0] * self.length
        self.stack, self.is_in_stack = [], [False] * self.length
        self.cnt, self.number = 0, 0
        self.ans = []


    def tarjan_(self, u:int):
        self.cnt += 1
        self.stack.append(u)
        self.is_in_stack[u] = True
        self.low[u] = self.dfn[u] = self.cnt
        for v in range(self.length):
            if self.g[u][v]:
                if not self.dfn[v]:
                    self.tarjan_(v)
                    self.low[u] = min(self.low[u], self.low[v])
                elif self.is_in_stack[v]:
                    self.low[u] = min(self.low[u], self.dfn[v])
        if self.dfn[u] == self.low[u]:
            self.number += 1
            tmp = set()
            j = -1
            while j != u:
                j = self.stack[-1]
                tmp.add(j + 1)
                self.stack.pop()
                self.is_in_stack[j] = False
            self.ans.append(tmp)


    def run_tarjan(self):
        self.tarjan_(0)
        for s in self.ans:
            print(s)


if __name__ == '__main__':
    gg = [[0, 1, 1, 0], [0, 0, 0, 0], [0, 1, 0, 1], [1, 0, 0, 0]]
    g = [[0, 1, 1, 0, 0, 0], [0, 0, 0, 1, 0, 0], [0, 0, 0, 1, 1, 0],
         [1, 0, 0, 0, 0, 1], [0, 0, 0, 0, 0, 1], [0, 0, 0, 0, 0, 0]]
    tar = Tarjan(g)
    tar.run_tarjan()